### PR TITLE
More work towards building ports on serenity

### PR DESCRIPTION
**Pull request by petelliott**
- fix .port_include.sh on serenity
- make fgrep take file arguments by making it part of grep
- add flags to sort
- fix printf's %g

I still haven't got it working but I figure more POSIX compatibility doesn't hurt

### CODE REVIEW

File: Ports/.port_include.sh [hunk 1]
```diff
@@ -41,7 +41,8 @@ target_env() {
 
 target_env
 
-DESTDIR="${SERENITY_INSTALL_ROOT}"
+SERENITY_ARCH="${SERENITY_ARCH:-$(uname -m)}"
+DESTDIR="${SERENITY_INSTALL_ROOT:-}"
```
Comment (timschumi) [inline Ports/.port_include.sh:45]: Will these even work, considering that we just sourced `.hosted_defs.sh` in the previous line?

In fact, both of these settings should just be in a special case inside `.hosted_defs.sh`.

File: Userland/Utilities/sort.cpp [hunk 2]
```diff
@@ -1,45 +1,126 @@
...
+
+        return l.key.hash();
+    }
+};
+
+struct Options {
+    size_t key_field { 0 };
+    bool unique { false };
+    bool numeric { false };
+    StringView seperator { "\0", 1 };
```
Comment (timschumi) [inline Userland/Utilities/sort.cpp:57]: This is a mistake that I often make as well, but it turns out that the correct spelling is "separator".

**Review by timschumi — CHANGES_REQUESTED** [review]

File: Userland/Utilities/sort.cpp (hunk 2 above)
Comment (timschumi) [reply Userland/Utilities/sort.cpp:57]: Also, couldn't one just initialize from `"\0"sv` here? The constructor that constructs from a string literal shouldn't be using `strlen`, right?

File: Ports/.port_include.sh (hunk 1 above)
Comment (petelliott) [reply Ports/.port_include.sh:45]: `.hosted_defs.sh` isn't copied to serenity. it contains settings that you only want when cross compiling.

File: Userland/Utilities/sort.cpp (hunk 2 above)
Comment (petelliott) [reply Userland/Utilities/sort.cpp]: both done

File: Ports/.port_include.sh (hunk 1 above)
Comment (timschumi) [reply Ports/.port_include.sh:45]: Could we at least do something like the following then? Setting the correct settings for on-Serenity builds via fallbacks _somewhere_ in the script doesn't feel quite right.

```diff
diff --git a/Ports/.port_include.sh b/Ports/.port_include.sh
index bf9e58a009..03d5844340 100755
--- a/Ports/.port_include.sh
+++ b/Ports/.port_include.sh
@@ -29,14 +29,16 @@ buildstep_intro() {
     echo -e "\x1b[1;32m=> $@\x1b[0m"
 }
 
-maybe_source() {
-    if [ -f "$1" ]; then
-        . "$1"
-    fi
-}
-
 target_env() {
-    maybe_source "${SCRIPT}/.hosted_defs.sh"
+    if [ -f "${SCRIPT}/.hosted_defs.sh" ]; then
+        . "${SCRIPT}/.hosted_defs.sh"
+    elif [ "$(uname -s)" = "SerenityOS" ]; then
+        export SERENITY_ARCH="$(uname -m)"
+        export SERENITY_INSTALL_ROOT=""
+    else
+        >&2 echo "Error: .hosted_defs.sh is missing and we are not running on Serenity."
+        exit 1
+    fi
 }
 
 target_env
```

Comment (timschumi) [reply Ports/.port_include.sh:45]: We'd presumably also want to do a similar check for `host_env`, which relies on variables set by `.hosted_defs.sh`.

Comment (petelliott) [reply Ports/.port_include.sh]: done. host_env looks unused though.

Comment (timschumi) [reply Ports/.port_include.sh]: It is only used by ports that need to build on the host, for example `llvm`.

**Comment by DanShaders** [comment]
As you are fixing `printf(double)`, it would be nice if you make it use `AK/StringFloatingPointConversions.h` when no precision is set.

**Review by timschumi — CHANGES_REQUESTED** [review]
The changes to `.ports_include.sh` I can't currently test, the rest looks and works fine apart from a few nitpicks.

**Comment by awesomekling** [comment]
This is failing the `clang-format` check.

File: Userland/Libraries/LibRegex/RegexParser.cpp [hunk 3]
```diff
@@ -2735,4 +2735,15 @@ size_t ECMA262Parser::ensure_total_number_of_capturing_parenthesis()
...
     return count;
 }
+
+bool FixedParser::parse_internal(ByteCode& bytecode, size_t& match_length_minimum)
+{
+    auto& string = m_parser_state.lexer.source();
+    match_length_minimum = string.length();
+    bytecode.insert_bytecode_compare_string(string);
+
+    // Parser::parse() expects us to be at eof.
```
[the remark this answers is not in the document]
Comment (petelliott) [reply Userland/Libraries/LibRegex/RegexParser.cpp:2745]: The EOF check happens here:

https://github.com/SerenityOS/serenity/blob/eb44a90e628be066dbfe685e19d3b64f977dcef1/Userland/Libraries/LibRegex/RegexParser.cpp#L185

File: AK/PrintfImplementation.h [hunk 4]
```diff
@@ -164,7 +164,7 @@ ALWAYS_INLINE int print_decimal(PutChFunc putch, CharType*& bufptr, u64 number,
 }
 #ifndef KERNEL
 template<typename PutChFunc, typename CharType>
-ALWAYS_INLINE int print_double(PutChFunc putch, CharType*& bufptr, double number, bool always_sign, bool left_pad, bool zero_pad, u32 field_width, u32 precision)
+ALWAYS_INLINE int print_double(PutChFunc putch, CharType*& bufptr, double number, bool always_sign, bool left_pad, bool zero_pad, u32 field_width, u32 precision, bool trailing_zeros)
```
[the remark this answers is not in the document]
Comment (petelliott) [reply AK/PrintfImplementation.h:167]: added

File: Userland/Libraries/LibRegex/RegexParser.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (timschumi) [reply Userland/Libraries/LibRegex/RegexParser.cpp]: Oh, I see, it checks for the EOF of the pattern, not for an EOF in the input that should be matched?

[the remark this answers is not in the document]
Comment (petelliott) [reply Userland/Libraries/LibRegex/RegexParser.cpp]: yes

**Review by linusg — CHANGES_REQUESTED** [review]
Typo in `FixedParser` commit: seperate -> separate

Seems like a strange solution to the problem though, we should probably look into escaping special regex characters in strings instead and then use the default parser 🤔

**Comment by linusg** [comment]
(also has conflicts)

File: Userland/Utilities/grep.cpp [hunk 5]
```diff
@@ -293,6 +296,14 @@ ErrorOr<int> serenity_main(Main::Arguments args)
         return did_match_something ? 0 : 1;
     };
 
+    if (fixed_strings) {
+        Vector<Regex<Fixed>> regular_expressions;
+        for (auto pattern : patterns) {
```
Comment (alimpfard) [inline Userland/Utilities/grep.cpp:301]: This is both smart and weird tbh, feels like regex shouldn't even be involved...

**Comment by AtkinsSJ** [comment]
This has conflicts.

**Comment by petelliott** [comment]
This should have fixed everything, except for the LibRegex thing. which I will fix when I get time next.

**Comment by petelliott** [comment]
> Typo in `FixedParser` commit: seperate -> separate
> 
> Seems like a strange solution to the problem though, we should probably look into escaping special regex characters in strings instead and then use the default parser thinking

this is now how it works.

**Comment by timschumi** [comment]
Would it make sense to force the selection of arguments to "basic regex" so that we only have to provide one "implementation" of unescaping characters? Realistically, `-E` and `-F` should be mutually exclusive anyways.

**Comment by petelliott** [comment]
> Would it make sense to force the selection of arguments to "basic regex" so that we only have to provide one "implementation" of unescaping characters? Realistically, `-E` and `-F` should be mutually exclusive anyways.

realistically I would have implemented it the same way either way, so removing the ere escapes would save a line of code and a string literal. realistically it doesn't matter so i figured why not give people running egrep the ere regex engine.

**Review by linusg — CHANGES_REQUESTED** [review]
CI is complaining about some man page thingy

**Comment by ADKaster** [comment]
Looks like this one is about to meet stalebot. @petelliott  can we get a rebase on this? Seems like the comments were all resolved.

**Comment by petelliott** [comment]
> CI is complaining about some man page thingy

fixed

File: Ports/.port_include.sh [hunk 6]
```diff
@@ -37,7 +37,15 @@ maybe_source() {
 }
 
 target_env() {
-    maybe_source "${SCRIPT}/.hosted_defs.sh"
```
Comment (GMTA) [inline Ports/.port_include.sh:40]: The function `maybe_source` is now unused and can be removed

**Pull request merged**